# SECOND INTERMEDIATE PYTHON SESSION
#### Written by Ioanna (Ioni) Kalospyrou for Women in Engineering x Women in Computer Science Python Workshop Series (2023)

## Introduction to Pandas
Pandas is a fast, powerful and easy to use open source data analysis and manipulation tool. it has functions for analysing, cleaning, exploring and manipulating data.

## Installing and importing Pandas
(More information on package installation has been provided in Intermediate Session 1)

In [20]:
!pip install pandas
import pandas as pd


## Example 1: Wine Dataset

### 1. Loading Data

Scikit learn is a free software machine learning library. It contains some toy data sets and for this tutorial we will choose to use the "Wine Recognition Dataset", which can be accessed by importing "load_wine" from sklearn.datasets as shown below:

In [21]:
from sklearn.datasets import load_wine

Then, you need to allocate the loaded wine data into a variable named "wine" and use this variable to define a Pandas DataFrame. A DataFrame is a data structure that organizes data into a 2-dimensional table of rows and columns, similar to a speadsheet.

In [22]:
wine = load_wine()
# A dataframe is usually symbolised with df, although, you can give it any name you wish.
df = pd.DataFrame(wine.data, columns=wine.feature_names)

EXTRA INFORMATION 

Data loading from computer file. 
You can open a spreadsheet from your computer and turn it into a DataFrame by following the following procedure:
1. Ensure that the spreadsheet is in the same folder to the one you have saved your code. 
2. Save the original spreadsheet in CSV (Comma Separated Values) format.
3. Create the Dataframe by writing the code: df = pd.read_csv('name_of_file.csv')

### 2. Data Inspection

To inspect the first lines of data saved in the DataFrame the code used is: df.head()
To inspect the last lines of data saved in the DataFrame the code used is: df.tail()

If left like this, the first or last 5 lines of data are generated respectively.
You can define the number of lines you want inspected in the parenthesis.

#### *Questions:*

a. Inspect the first 5 lines of data.

b. Inspect the last 10 lines of data.

c. Inspect the first 3 lines of data.

d. Inspect the last 2 lines of data.

In [23]:
# Answer Question a here:
print(df.head())

# Answer Question b here:
print(df.head(10))

# Answer Question c here:
print(df.tail(3))

# Answer Question d here:
print(df.tail(2))


   alcohol  malic_acid   ash  alcalinity_of_ash  magnesium  total_phenols  \
0    14.23        1.71  2.43               15.6      127.0           2.80   
1    13.20        1.78  2.14               11.2      100.0           2.65   
2    13.16        2.36  2.67               18.6      101.0           2.80   
3    14.37        1.95  2.50               16.8      113.0           3.85   
4    13.24        2.59  2.87               21.0      118.0           2.80   

   flavanoids  nonflavanoid_phenols  proanthocyanins  color_intensity   hue  \
0        3.06                  0.28             2.29             5.64  1.04   
1        2.76                  0.26             1.28             4.38  1.05   
2        3.24                  0.30             2.81             5.68  1.03   
3        3.49                  0.24             2.18             7.80  0.86   
4        2.69                  0.39             1.82             4.32  1.04   

   od280/od315_of_diluted_wines  proline  
0                  

### 3. Data Cleaning

You can delete a column or row from a DataFrame using the code df.drop(). In the parenthesis you should have the name of the column or row you are targeting as well as the axis. For columns, axis=1, for rows, axis=0.

#### *Question:*

Generate a new dataset, df1 which has all the information of df just not the "magnesium" and "malic_acid" columns.
Proceed to inspect the first 5 lines of df1.

In [24]:
#Answer here: 
df1 = df.drop('magnesium', axis=1)
df1 = df1.drop('malic_acid',axis=1)

print(df1.head())

   alcohol   ash  alcalinity_of_ash  total_phenols  flavanoids  \
0    14.23  2.43               15.6           2.80        3.06   
1    13.20  2.14               11.2           2.65        2.76   
2    13.16  2.67               18.6           2.80        3.24   
3    14.37  2.50               16.8           3.85        3.49   
4    13.24  2.87               21.0           2.80        2.69   

   nonflavanoid_phenols  proanthocyanins  color_intensity   hue  \
0                  0.28             2.29             5.64  1.04   
1                  0.26             1.28             4.38  1.05   
2                  0.30             2.81             5.68  1.03   
3                  0.24             2.18             7.80  0.86   
4                  0.39             1.82             4.32  1.04   

   od280/od315_of_diluted_wines  proline  
0                          3.92   1065.0  
1                          3.40   1050.0  
2                          3.17   1185.0  
3                          3

### 4. Statistics

You can obtain the statistical summaries for all columns of the DataFrame with the code:  df.describe()

#### *Questions:*

a. Print the statistical summary of the wine dataframe.

b. Print the statistical summary only of the "alcohol" column.

c. Print the statistical summary only of the "ash" and "proline" columns.


In [25]:
#Answer Question a here:
print(df.describe())
#once you answer this question, # it before you move on to the next to 
#have a clearer view of what you print.

#Answer Question b here:
print(df['alcohol'].describe())

#Answer Question c here:
print(df[['ash','proline']].describe())

          alcohol  malic_acid         ash  alcalinity_of_ash   magnesium  \
count  178.000000  178.000000  178.000000         178.000000  178.000000   
mean    13.000618    2.336348    2.366517          19.494944   99.741573   
std      0.811827    1.117146    0.274344           3.339564   14.282484   
min     11.030000    0.740000    1.360000          10.600000   70.000000   
25%     12.362500    1.602500    2.210000          17.200000   88.000000   
50%     13.050000    1.865000    2.360000          19.500000   98.000000   
75%     13.677500    3.082500    2.557500          21.500000  107.000000   
max     14.830000    5.800000    3.230000          30.000000  162.000000   

       total_phenols  flavanoids  nonflavanoid_phenols  proanthocyanins  \
count     178.000000  178.000000            178.000000       178.000000   
mean        2.295112    2.029270              0.361854         1.590899   
std         0.625851    0.998859              0.124453         0.572359   
min         0.9

You can also obtain specific statistic information on different columns with .count(), 
.mean(), .std(), .min(), .max()

#### *Questions:*

a. Print the mean of all columns in the wine dataframe.

b. Which has a higher standard deviation? Malic acid or proline?

c. Print the minimum only of the "ash","proanthocyanins" and "alcohol" columns.



In [26]:
#Answer Question a here:
print(df.mean())
#once you answer this question, # it before you move on to the next, to have a clearer
#view of what you print.

#Answer Question b here:
print(df[['malic_acid','proline']].std())
# The answer is proline

#Answer Question c here:
print(df[['ash','proanthocyanins','alcohol']].min())

alcohol                          13.000618
malic_acid                        2.336348
ash                               2.366517
alcalinity_of_ash                19.494944
magnesium                        99.741573
total_phenols                     2.295112
flavanoids                        2.029270
nonflavanoid_phenols              0.361854
proanthocyanins                   1.590899
color_intensity                   5.058090
hue                               0.957449
od280/od315_of_diluted_wines      2.611685
proline                         746.893258
dtype: float64
malic_acid      1.117146
proline       314.907474
dtype: float64
ash                 1.36
proanthocyanins     0.41
alcohol            11.03
dtype: float64


### 5. Plotting (with Plotly Express)

The plotly Python library is an interactive, open-source plotting library. The plots generated using this library are easy to make and much more elegant that the ones generated with Matplotlib. 
Extra info: The top library for plots and visualisations is Seaborn. 

It is installed with the following code:

In [27]:
!pip install plotly==5.13.1
import plotly.express as px

#### Two dimensional Scatter Plot using Plotly Express



The code used to generate a figure, namely, a plot with plotly express is the following:

fig = px.scatter(df, x, y)

fig.show()

where df is the dataframe you wish to define, x needs to be defined as the dataframe column you will assign the x axis to and y is the dataframe column you will assign the y axis to.
Never forget the .show() code at the end for your figure to be generated. If you don't include it, you will not be able to see the plot (it will be generated though regardless).

You can generate multiple types of plots by changing the .scatter accordingly.

For a line plot: .line()

For a histogram: .histogram()

For a bar chart: .bar()

For a pie chart: .pie()

of course, the arguments within the brackets change since different charts have different requirements. For more info: https://plotly.com/python/

As engineers though, more often than not, we have to deal with ranges of experimental data points, where trends and correlations are important to be understood. Hence, we will be focusing on scatter plots.

#### *Question:*

Generate a scatter plot of alcohol against malic acid.


In [28]:
# Answer here
fig = px.scatter(df,x='malic_acid',y='alcohol')
fig.show()

#### 2D multivariate scatter plotting

Scatter plots with variable-sized circular markers are often known as bubble charts. 
By adding the arguments, color and size in the scatter brackets, different variables can affect those. Consequenctly, multivariate relationships can be represented. 

##### *Question:*

Create a scatter plot of alcohol against malic acid, where the datapoint size will be representative of the proline and the color of the magnesium.

In [29]:
# Answer here
fig = px.scatter(df,x='malic_acid',y='alcohol',color = 'magnesium', size = 'proline')
fig.show()

## Example 2: Open Power System Data - Time Series Data

### 1. Loading Data

Open Power System Data is a publicly available dataset containing electricity consumption, wind power production, and solar power production data for 2006-2017. The data is loaded below:

In [30]:
url='https://raw.githubusercontent.com/jenfly/opsd/master/opsd_germany_daily.csv'
df = pd.read_csv(url,sep=",")

#### *Question:*

View the whole dataset df.

In [31]:
#Answer here:
print(df)

            Date  Consumption     Wind   Solar  Wind+Solar
0     2006-01-01   1069.18400      NaN     NaN         NaN
1     2006-01-02   1380.52100      NaN     NaN         NaN
2     2006-01-03   1442.53300      NaN     NaN         NaN
3     2006-01-04   1457.21700      NaN     NaN         NaN
4     2006-01-05   1477.13100      NaN     NaN         NaN
...          ...          ...      ...     ...         ...
4378  2017-12-27   1263.94091  394.507  16.530     411.037
4379  2017-12-28   1299.86398  506.424  14.162     520.586
4380  2017-12-29   1295.08753  584.277  29.854     614.131
4381  2017-12-30   1215.44897  721.247   7.467     728.714
4382  2017-12-31   1107.11488  721.176  19.980     741.156

[4383 rows x 5 columns]


As we can see from df, the index is 0 to 4383. We would like to change the index to the date. To change the index apply the code: df=df.set_index('column_of_interest')


#### *Question*:

Change the index column to Date. View the first 5 lines of the df after you set the new index.

In [13]:
#Answer here
df = df.set_index('Date')
print(df.head())

            Consumption  Wind  Solar  Wind+Solar
Date                                            
2006-01-01     1069.184   NaN    NaN         NaN
2006-01-02     1380.521   NaN    NaN         NaN
2006-01-03     1442.533   NaN    NaN         NaN
2006-01-04     1457.217   NaN    NaN         NaN
2006-01-05     1477.131   NaN    NaN         NaN


### 3. Data Preprocessing/ Data Cleaning

Data pre-processing is the processing that needs to be performed on the raw data to transform it into a more effective and understandable format. The most common problems that occur during this process are outliers, missing values and duplicated data. In this particular data set, all data points refer to verified values hence outliers will not be considered in the analysis and no duplicates are present.

#### How to deal with missing values

There are multiple ways to deal with missing values:

1. Eliminate rows with missing values - dropna (axis=0)
2. Forward fill: Fill values with preceding data – fillna (method = ‘ffill’) 
3. Backward fill: Fill values with following data – fillna (method = ‘bfill’) 
4. Linear Interpolation: Fill values using linear interpolation – interpolate (method = ‘linear’) 

It is important to ensure that filling of data will not add bias and trends to the dataframe. The case where we can use filling methods with a greater level of certainty that there will be no bias is when dealing with time series data, like the manipulated dataframe, "Open Power System Data".

#### *Question:*

Generate a new dataframe, df_new, which is equal to the original df with all rows with NaN values eliminated. Print the shape of df and df_new using .shape to check if the rows with NaN values were actually eliminated. 

In [14]:
#Answer here:
df_new = df.dropna(axis = 0)
print(df.shape) 
print(df_new.shape)

(4383, 4)
(2187, 4)


#### *Question:*

Generate dataframes df_ff and df_bf, where NaN values are filled with forward and backward fill respectively. Compare the means of both dataframes (Hint: Look back to Example 1, Section 4). 


In [15]:
df_ff = df.fillna(method = 'ffill')
df_bf = df.fillna(method = 'bfill')
print(df_ff.mean())
print(df_bf.mean())



Consumption    1338.675836
Wind            164.884228
Solar            89.265476
Wind+Solar      272.461183
dtype: float64
Consumption    1338.675836
Wind            126.107998
Solar            47.960698
Wind+Solar      253.235085
dtype: float64
